api/models.py

```diff
@@ -215,6 +215,19 @@ def set_timeout(self, hours: int = 24):
     def update(self):
         self.updated = datetime.utcnow()
 
+    @classmethod
```

Comment:
Is that not done already by pydantic?

Reply:
Sounds like the same question @mgalka asked a while ago, and I still don't understand the rationale for not using pydantic's standard datetime validation.

Reply:
As I already explained, pydantic validators defined with `validator` decorator will be invoked when an object of a model is instantiated. It can be used while creating or updating documents e.g.
```
try:
    Node(name="kunit-test-case", path=["checkout","kunit","kunit-test-case"],...)
except ValidationError as e:
    print(e)
```
On the other hand, `validate_params` is defined to validate request query parameters received with GET request.
These request parameters are being used as attributes to get required documents from the database and not to create any documents.
So, this is the validation of request query params and not object fields.

Reply:
When receiving a request in `main.py`, and the handler function specifies a pydantic model, then validation will be made on the incoming data.  For example:
```
@app.post('/node', response_model=Node)
async def post_node(node: Node, token: str = Depends(get_user)):
```
The input data will need to produce a valid `Node` object, if the pydantic validation doesn't accept it then an HTTP response is directly sent to the client.

Reply:
I think the issue here is that some use-case from the endpoint handler is being "leaked" into the model.  The model shouldn't know about how it's being used in other parts of the code.  In any case, there's some code in pydantic to validate fields such as timestamps and I still don't see any good reason why we would not be using it.  The actual problem is how to validate parameters for GET requests, and maybe that should be done outside the model.

Reply:
That's exactly what I am trying to say.
We can use the pydantic validators in POST and PUT requests but in this case, we need to validate request params manually.
I am not sure if we should move out this validation from the model, as these validations are model specific and would be different for every model.

Reply:
Yes, if we want to use [FastAPI validation](https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#add-more-validations) for query params, we need to mention all the params explicitly which is not feasible in our case as we have so many fields.

Reply:
Yes but pydantic is separate from FastAPI so we might be able to use pydantic validation directly on the fields.  I'm not sure, just a thought.

Reply:
Currently, data type validations are already being handled by default for POST and PUT requests.
I don't think we need any extra validations atm.

Reply:
I mean, doing something like `pydantic.validate(type=datetime, input_value)` (making things up here...) could avoid us having to write some custom validation logic.  But I think we can park that for now, it's not a blocker for this PR.